src/hotspot/share/opto/superword.cpp

```diff
@@ -2877,10 +2883,139 @@ bool SuperWord::output() {
   }
+
+  for (uint i = 0; i < unordered_reductions.size(); i++) {
+    UnorderedReductionNode* ur = unordered_reductions.at(i)->as_UnorderedReduction();
+    move_unordered_reduction_out_of_loop(ur);
+  }
+
   NOT_PRODUCT(if(is_trace_loop_reverse()) {tty->print_cr("\n Final loop after SuperWord"); print_loop(true);})
   return true;
 }
 
+// Having a ReductionNode in the loop is expensive. It needs to recursively
+// fold together the vector values, for every vectorized loop iteration. If
+// we encounter the following pattern, we can move the UnorderedReduction
+// outside the loop.
+//
+// CountedLoop     init
+//          |        |
+//          +------+ | +---------------+
+//                 | | |               |
+//                PhiNode (s)  Vector  |
+//                  |            |     |
+//               UnorderedReduction    |
+//                       |             |
+//                       +-------------+
+//
+// We patch the graph to look like this:
+//
+// CountedLoop   neutral_vector
+//         |         |
+//         +-------+ | +---------------+
+//                 | | |               |
+//                PhiNode (v)  Vector  |
+//                   |           |     |
+//      init       VectorAccumulator   |
+//        |          |     |           |
+//     UnorderedReduction  +-----------+
+//
+// We turned the scalar (s) Phi into a vectorized one (v). In the loop, we
+// use a vector_accumulator, which does the same reduction, but only element
+// wise. This is a single operation, rather than many for the ReductionNode.
+// We can then reduce that vector_accumulator after the loop, and also reduce
+// the init value into it.
+// We can not do this with all reductions. Some reductions do not allow the
+// reordering of operations (for example float addition).
+void SuperWord::move_unordered_reduction_out_of_loop(UnorderedReductionNode* ur) {
+  Compile* C = _phase->C;
+  CountedLoopNode *cl = lpt()->_head->as_CountedLoop();
+  Node* ctrl = ur->in(0);
+  Node* in1  = ur->in(1);
+  Node* in2  = ur->in(2);
+
+  // Expect no ctrl for ur
+  if (ctrl != nullptr) {
+    return;
+  }
+
+  // Expect data loop over backedge of Phi in cl
+  if (in1 == nullptr || !in1->is_Phi() || in1->in(2) != ur || in1->outcnt() != 1 ||
+      in1->in(0) != cl) {
+    return;
+  }
+
+  // Expect all uses to be outside the loop, except Phi
+  PhiNode* phi = in1->as_Phi();
+  for (DUIterator_Fast kmax, k = ur->fast_outs(kmax); k < kmax; k++) {
+    Node* use = ur->fast_out(k);
+    if (use != phi && _phase->ctrl_or_self(use) == cl) {
+      return;
+    }
+  }
+
+  // Expect input vector inside loop
+  if (!in2->is_Vector() || _phase->get_ctrl(in2) != cl) {
+    return;
+  }
+  VectorNode* vector = in2->as_Vector();
+
+  // Determine types
+  BasicType bt = ur->vect_type()->element_basic_type();
+  const Type* bt_t = Type::get_const_basic_type(bt);
+
+  // Create vector of neutral elements (zero for add, one for mul, etc)
```

Comment:
@jatin-bhateja Ok, "neutral element" and "identity element" seem to be synonyms. I'll change it to "identity", since that is what we seem to use in the code already.